pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java

```diff
@@ -226,50 +250,115 @@ private void handleMessage(Message<T> msg) {
+                                   Message<T> msg) {
+        // The message received from multi-consumer/multi-reader is processed to TopicMessageImpl.
+        String topicName = msg.getTopicName();
+        TopicMessageId maxMessageId = maxMessageIds.get(topicName);
+        // We need remove the partition from the maxMessageIds map
+        // once the partition has been read completely.
+        if (maxMessageId != null && msg.getMessageId().compareTo(maxMessageId) >= 0) {
+            maxMessageIds.remove(topicName);
+        }
+        if (maxMessageIds.isEmpty()) {
+            future.complete(null);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private void checkAllFreshTask(Message<T> msg) {
+        refreshRequests.forEach((future, maxMessageIds) -> {
+            if (checkFreshTask(maxMessageIds, future, msg)) {
+                refreshRequests.remove(future);
+            }
+        });
+    }
+
+    private void readAllExistingMessages(Reader<T> reader, CompletableFuture<Void> future, long startTime,
                                          AtomicLong messagesRead, Map<String, TopicMessageId> maxMessageIds) {
         reader.hasMessageAvailableAsync()
                 .thenAccept(hasMessage -> {
                    if (hasMessage) {
                        reader.readNextAsync()
                                .thenAccept(msg -> {
                                   messagesRead.incrementAndGet();
-                                  // We need remove the partition from the maxMessageIds map
-                                  // once the partition has been read completely.
-                                  TopicMessageId maxMessageId = maxMessageIds.get(msg.getTopicName());
-                                  if (maxMessageId != null && msg.getMessageId().compareTo(maxMessageId) >= 0) {
-                                      maxMessageIds.remove(msg.getTopicName());
-                                  }
-                                  handleMessage(msg);
-                                  if (maxMessageIds.isEmpty()) {
-                                      future.complete(reader);
-                                  } else {
-                                      readAllExistingMessages(reader, future, startTime, messagesRead, maxMessageIds);
-                                  }
+                                   try {
+                                       handleMessage(msg);
+                                       if (!checkFreshTask(maxMessageIds, future, msg)) {
+                                           readAllExistingMessages(reader, future, startTime,
+                                                   messagesRead, maxMessageIds);
+                                       }
+                                   } finally {
+                                       msg.release();
+                                   }
```

Comment:
Sorry my [previous comment](https://github.com/apache/pulsar/pull/21417#discussion_r1519288345) ignores the case here. 

`handleMessage` will release the reference count of `msg`, so it's unsafe to pass the `msg` to `checkFreshTask` after that. You might get around it like:

```java
String topicName = msg.getTopicName();
MessageId msgId = msg.getMessageId();
handleMessage(msg);
if (!checkFreshTask(maxMessageIds, future, topicName, msgId)) {
```